example/poisson/poisson_package.cpp

```diff
@@ -49,59 +50,173 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
+    pkg->AddParam("sparse_accessor", sp_accessor);
+  }
+
   return pkg;
 }
 
+template <typename T>
+TaskStatus SetMatrixElements(T *u) {
+  auto pm = u->GetParentPointer();
+
+  IndexRange ib = u->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = u->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = u->GetBoundsK(IndexDomain::interior);
+
+  PackIndexMap imap;
+  const std::vector<std::string> vars({"poisson_sparse_matrix"});
+  const auto &v = u->PackVariables(vars, imap);
+  const int isp_lo = imap["poisson_sparse_matrix"].first;
+  const int isp_hi = imap["poisson_sparse_matrix"].second;
+
+  if (isp_hi < 0) { // must be using the stencil so return
+    return TaskStatus::complete;
+  }
+
+  const int ndim = v.GetNdim();
+  const Real w0 = 1.0 / (2.0 * ndim);
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "SetMatElem", DevExecSpace(), 0, v.GetDim(5) - 1, kb.s, kb.e,
+      jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        for (int n = isp_lo; n <= isp_hi; n++) {
+          v(b, n, k, j, i) = w0;
+        }
+        v(b, isp_lo + 1, k, j, i) = -1.0;
+      });
+
+  return TaskStatus::complete;
+}
+
 auto &GetCoords(std::shared_ptr<MeshBlock> &pmb) { return pmb->coords; }
 auto &GetCoords(Mesh *pm) { return pm->block_list[0]->coords; }
 
 template <typename T>
-TaskStatus UpdatePhi(T *u, T *du) {
-  Kokkos::Profiling::pushRegion("Task_Poisson_UpdatePhi");
+TaskStatus SumMass(T *u, Real *reduce_sum) {
   auto pm = u->GetParentPointer();
 
   IndexRange ib = u->GetBoundsI(IndexDomain::interior);
   IndexRange jb = u->GetBoundsJ(IndexDomain::interior);
   IndexRange kb = u->GetBoundsK(IndexDomain::interior);
 
   PackIndexMap imap;
-  const std::vector<std::string> vars({"density", "potential"});
+  const std::vector<std::string> vars({"density"});
   const auto &v = u->PackVariables(vars, imap);
   const int irho = imap["density"].first;
-  const int iphi = imap["potential"].first;
-  const std::vector<std::string> phi_var({"potential"});
-  PackIndexMap imap2;
-  const auto &dv = du->PackVariables(phi_var, imap2);
-  const int idphi = imap2["potential"].first;
 
   auto coords = GetCoords(pm);
   const int ndim = v.GetNdim();
   const Real dx = coords.Dx(X1DIR);
   for (int i = X2DIR; i <= ndim; i++) {
     const Real dy = coords.Dx(i);
     PARTHENON_REQUIRE_THROWS(dx == dy,
-                             "UpdatePhi requires that DX be equal in all directions.");
+                             "SumMass requires that DX be equal in all directions.");
+  }
+
+  Real total;
+  parthenon::par_reduce(
+      parthenon::loop_pattern_mdrange_tag, "SumMass", DevExecSpace(), 0, v.GetDim(5) - 1,
+      kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &sum) {
+        sum += v(b, irho, k, j, i) * std::pow(dx, ndim);
+      },
+      Kokkos::Sum<Real>(total));
```

Comment:
So for my reduction needs, my equivalent local kernel reduces two values at once and is somewhat expensive to execute. I don't want to execute the kernel twice, so I don't want two tasks for the local reduction. With the current interface, I think I could still accomplish the local reduction all at once into two different `AllReduce` objects. It would then be only a little awkward to add separate global reduction tasks for the two `AllReduce` objects.